extensions/policies/src/main/java/org/eclipse/dataspaceconnector/mvd/RegionConstraintFunction.java

```
@@ -0,0 +1,81 @@
+
+public class RegionConstraintFunction implements AtomicConstraintFunction<Permission> {
+    private static final String REGION_KEY = "region";
+    private final ObjectMapper objectMapper;
+    private final Monitor monitor;
+
+    public RegionConstraintFunction(ObjectMapper objectMapper, Monitor monitor) {
+        this.objectMapper = objectMapper;
+        this.monitor = monitor;
+    }
+
+    @Override
+    public boolean evaluate(Operator operator, Object rightValue, Permission rule, PolicyContext context) {
+        var regions = getRegions(context.getParticipantAgent().getClaims());
+        switch (operator) {
+            case EQ:
+                return regions.contains(rightValue);
+            case NEQ:
+                return !regions.contains(rightValue);
+            default:
+                return false;
+        }
+    }
+
+    private List<String> getRegions(Map<String, Object> claims) {
+        return claims.values()
+                .stream().flatMap(o -> getVerifiableCredential(o).stream())
+                .flatMap(vc -> getRegion(vc).stream()).collect(Collectors.toList());
+    }
+
+    private Optional<VerifiableCredential> getVerifiableCredential(Object object) {
+        try {
+            var vcObject = (Map<String, Object>) object;
+            var verifiableCredentialMap = vcObject.get(VERIFIABLE_CREDENTIALS_KEY);
+            return Optional.of(objectMapper.convertValue(verifiableCredentialMap, VerifiableCredential.class));
+        } catch (ClassCastException e) {
+            monitor.warning("Error getting verifiable credentials", e);
+            return Optional.empty();
+        }
+    }
+
+    private Optional<String> getRegion(VerifiableCredential verifiableCredential) {
+        try {
+            var region = verifiableCredential.getCredentialSubject().get(REGION_KEY);
+            return Optional.ofNullable((String) region);
+        } catch (Exception e) {
```

nit: is there a particular reason why are catching all of the exception here and then returning empty optional, instead not letting exception be thrown or bubble up?

same as in above method `getVerifiableCredential`

---

I changed the exception to ClassCastException to make it better.
IT is in case a region can't be casted to string. To avoid breaking the entire flow checking the policy.

---

or maybe I was also thinking too much your trick with optional may also work if you catch all exception with logging error about it.

And just a java doc on the method that we continue to process to avoid any Verifiable credential which do not have region as per the correct format.